Support for weak binding views to allow dropping of underlying objects

R4hul04 opened the pull request:
### Description

This PR introduces improvements to Babelfish's view handling, focusing on weak binding views. The changes enable more SQL Server-compatible behavior for view dependencies.

This PR introduces the following changes:
1. Added hook in DROP codepath to allow dropping underlying objects (tables, functions, views) referenced by weak binding views without errors, while maintaining the view structure
2. Implemented a view repair mechanism hook that automatically fixes broken views when their referenced objects are recreated, eliminating the need for manual view recreation

Implementation Details
1. Added a new flags `BBF_VIEW_DEF_FLAG_IS_BROKEN` to track views with missing dependencies and `BBF_VIEW_DEF_FLAG_IS_WEAK_VIEW` to identify view's schemabinding property in `babelfish_view_def` catalog
2. Implemented hooks in object drop operations to detect and mark dependent views as broken
4. Added preQueryRewrtie hook for view repair mechanism that activates when a broken view is accessed during the rewriting
5. Wrapper function for DefineVirtualRelation
 
### Issues Resolved

BABEL-1660

Signed-off-by: Rahul Parande [rparande@amazon.com](mailto:rparande@amazon.com)
 
### Check List

- [ ] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is under the terms of the PostgreSQL license, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/postgresql_modified_for_babelfish/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

Commit:
Support for weak binding views to allow dropping of underlying objects

Signed-off-by: Rahul Parande <rparande@amazon.com>

Commit:
Added view repair hook definiton in rewriteHandler.h

Signed-off-by: Rahul Parande <rparande@amazon.com>

Commit:
Fixed indentation

Signed-off-by: Rahul Parande <rparande@amazon.com>

Commit:
Fix indentation in pg_dump.c

Signed-off-by: Rahul Parande <rparande@amazon.com>

Commit:
Fixed variable declaration in pg_dump.c

Signed-off-by: Rahul Parande <rparande@amazon.com>

forestkeeper commented on src/backend/commands/dropcmds.c:
```diff
@@ -131,6 +133,20 @@ RemoveObjects(DropStmt *stmt)
 
 		add_exact_object_address(&address, objects);
+
+		/* Check for strong views and handle weak views */
```
Why not move the open table into the hook implement ?

Commit:
Added hook in view creation path to check view references on strong
views

Signed-off-by: Rahul Parande <rparande@amazon.com>

Commit:
Added a dialect check when calling the hook in view.c

Signed-off-by: Rahul Parande <rparande@amazon.com>

Commit:
Removed dialect check from view.c and added inside hook

Signed-off-by: Rahul Parande <rparande@amazon.com>

Commit:
Moved code for opening table inside hook instead in engine

Signed-off-by: Rahul Parande <rparande@amazon.com>

forestkeeper commented on src/bin/pg_dump/pg_dump.c:
```diff
@@ -15822,6 +15822,50 @@ dumpTable(Archive *fout, const TableInfo *tbinfo)
+ * This returns a new buffer which must be freed by the caller.
+ */
+static PQExpBuffer
+createDummyViewAsClause(Archive *fout, const TableInfo *tbinfo)
```
Why not reuse the createDummyViewAsClause def from community ?

R4hul04 replied:
We are reusing the same function. We are only moving it before `createViewAsClause`, since we need  `createDummyViewAsClause` defined before the function call in `createViewAsClause`

forestkeeper replied:
We should not move this function's place in community code, since after moving, the follow up new commits diff from community will be lost for this function.

R4hul04 replied:
Agreed. 
Added the function declaration of `createDummyViewAsClause` at the top instead of moving the whole function body.

forestkeeper commented on src/bin/pg_dump/pg_dump.c:
```diff
@@ -15833,6 +15877,7 @@ createViewAsClause(Archive *fout, const TableInfo *tbinfo)
 {
 	PQExpBuffer query = createPQExpBuffer();
 	PQExpBuffer result = createPQExpBuffer();
+	PQExpBuffer dummyResult = createPQExpBuffer();
```
Dummyresult doesn't need to be created as createPQExpBuffer.

R4hul04 replied:
Removed `dummyResult` variable instead using the existing variable `result`

forestkeeper commented on src/bin/pg_dump/pg_dump.c:
```diff
@@ -15869,50 +15927,6 @@ createViewAsClause(Archive *fout, const TableInfo *tbinfo)
 }
 
-/*
- * Create a dummy AS clause for a view.  This is used when the real view
```
Why move the whole function definition ?

R4hul04 replied:
Added `createDummyViewAsClause` inside `createViewAsClause`. Hence we needed to define createDummyViewAsClause before createViewAsClause.

forestkeeper replied:
We should find a alternative way rather than moving this function.

R4hul04 replied:
Added the function declaration of `createDummyViewAsClause` at the top instead of moving the whole function body.

forestkeeper commented on src/backend/commands/dropcmds.c:
```diff
@@ -19,9 +19,11 @@
 #include "catalog/dependency.h"
 #include "catalog/namespace.h"
 #include "catalog/objectaddress.h"
+#include "catalog/pg_depend_d.h"
```
Why adding this .h ?

R4hul04 replied:
Removed pg_depend_d.h

Commit:
Removed unnecessary header file pg_depend.h from dropcmds.c file

Signed-off-by: Rahul Parande <rparande@amazon.com>

tanscorpio7 commented on src/backend/rewrite/rewriteHandler.c:
```diff
@@ -4431,6 +4431,14 @@ QueryRewrite(Query *parsetree)
+     */
+	if (view_repair_hook)
+		(*view_repair_hook)(parsetree);
+	/*
```
Should we use a more generic name. pre_query_rewrite_hook

kuntalghosh replied:
+1
Let's use the name `pre_QueryRewrite_hook`.

R4hul04 replied:
Renamed the hook to `pre_QueryRewrite_hook`

tanscorpio7 commented on src/backend/commands/tablecmds.c:
```diff
@@ -1676,6 +1677,13 @@ RemoveRelations(DropStmt *drop)
+		{			
+			if (!(*view_dependency_hook)(&obj, NULL, NULL))
+				continue;
+		}
```
Let us move all comments and conditions inside the hook. Also would naming it `pre_drop_object_hook` make sense ?

R4hul04 replied:
Moved implementation to `InvokeObjectDropHook` instead.

Commit:
Reverted the placement of createDummyViewAsClause to original and
instead added it's declaration at the top to access the function

Signed-off-by: Rahul Parande <rparande@amazon.com>

forestkeeper commented on src/backend/commands/tablecmds.c:
```diff
@@ -1676,6 +1677,13 @@ RemoveRelations(DropStmt *drop)
 		}
+
+		/* Check for strong views and handle weak views */
+		if ((drop->removeType == OBJECT_TABLE || drop->removeType == OBJECT_VIEW) && view_dependency_hook)
```
Can we move this part also into the hook : (drop->removeType == OBJECT_TABLE || drop->removeType == OBJECT_VIEW) ?

kuntalghosh replied:
Let's use InvokeObjectDropHook which already there in the function.

R4hul04 replied:
Moved implementation to use `InvokeObjectDropHook` instead.

kuntalghosh commented on src/backend/commands/dropcmds.c:
```diff
@@ -131,6 +132,13 @@ RemoveObjects(DropStmt *stmt)
+		{			
+			if (!(*view_dependency_hook)(&address, NULL, NULL))
+				continue;
+		}
```
Let's make the hook generic and add all TSQL related check within the hook function. We're not worried about hook overhead in a DDL execution. Also, there's already a hook for this task - InvokeObjectDropHook. Let's plugin this hook here.

R4hul04 replied:
Used `InvokeObjectDropHook` instead of `view_dependency_hook`.

kuntalghosh commented on src/backend/commands/view.c:
```diff
@@ -381,6 +384,9 @@ DefineView(ViewStmt *stmt, const char *queryString,
 
+	if (check_view_dependencies_hook)
+			(*check_view_dependencies_hook)(viewParse);
+
```
The function parse_analyze_fixedparams already has a hook inside it - post_parse_analyze_hook(). Let's use that.

R4hul04 replied:
Updated `post_parse_analyze_hook`() with the required logic.

kuntalghosh commented on src/bin/pg_dump/pg_dump.c:
```diff
@@ -15856,8 +15856,19 @@ createViewAsClause(Archive *fout, const TableInfo *tbinfo)
 	if (len == 0)
-		pg_fatal("definition of view \"%s\" appears to be empty (length zero)",
-				 tbinfo->dobj.name);
+	{
+		/* 
+		 * Handle broken views (with empty definitions)
+		 * Instead of failing, create a dummy SELECT that preserves the structure
+		 */
+		PQclear(res);
+		destroyPQExpBuffer(query);
+		
+		/* Use createDummyViewAsClause to generate a compatible structure */
+		result = createDummyViewAsClause(fout, tbinfo);
+		
+		return result;
+	}
```
Whenever we modify in pg_dump, we make sure that this is a Babelfish Database. Check the functions with `bbf` prefix. Ideally, the logic should be,
```
if (bbfShouldDumpViews(Archive *fout, const TableInfo *tbinfo))
{
    createTSQLViewAsClause()
    return;
}
```
Within `createTSQLViewAsClause`, we should do all the handlings.

R4hul04 replied:
Do we need to create our own `createTSQLViewAsClause` function in `dump_babel_utils.c` which will be called specifically for babelfish database. This function will be called from `createViewAsClause` in `pg_dump.c` ?

tanscorpio7 replied:
I thought we will insert a dummy view query into pg_rewrite at the time of DROP which will avoid the need to alter dump restore. The dummy query can be prepared according to the columns.
We should avoid keeping empty string as rewrite rule since engine code assumes that is an impossible case.

R4hul04 replied:
Reverted back to original code and inserted dummy_query in pg_rewrite during DROP, which takes care of pg_dump for broken views

Commit:
1. Replaced view_dependency_hook with pre-existing hook
InvokeObjectDropHook.
2. Removed check_view_dependencies_hook from view.c and used existing
   post_parse_analyze_hook.
3. Renamed view_repair_hook to pre_QueryRewrite_hook

Signed-off-by: Rahul Parande <rparande@amazon.com>

Commit:
Removed unnecessary hook declarations

Signed-off-by: Rahul Parande <rparande@amazon.com>

Commit:
Added dialect check for function drop before calling InvokeObjectDropHook

Signed-off-by: Rahul Parande <rparande@amazon.com>

Commit:
updated argument in InvokeObjectDropHook for subid as -1 to recognize it
as special call for dependency check during function drop

Signed-off-by: Rahul Parande <rparande@amazon.com>

Commit:
Reverted to original code in createViewAsClause

Signed-off-by: Rahul Parande <rparande@amazon.com>

Commit:
Removed unnecessary hook declarations and header files

Signed-off-by: Rahul Parande <rparande@amazon.com>

Commit:
Changed definition of DefineVirtualRelation from static to extern to
reuse the logic for weak binding view repair

Signed-off-by: Rahul Parande <rparande@amazon.com>

tanscorpio7 commented on src/bin/pg_dump/pg_dump.c:
```diff
@@ -345,7 +345,6 @@ static TableInfo *getRootTableInfo(const TableInfo *tbinfo);
 static bool forcePartitionRootLoad(const TableInfo *tbinfo);
 static void read_dump_filters(const char *filename, DumpOptions *dopt);
 
-
```
Let us make sure we do not commit any unnecessary changes, they will end up causing merge conflicts during community merge.

tanscorpio7 commented on src/include/commands/view.h:
```diff
@@ -27,4 +27,8 @@ extern PGDLLEXPORT store_view_definition_hook_type	store_view_definition_hook;
 
 typedef void (*inherit_view_constraints_from_table_hook_type) (ColumnDef  *col, Oid tableOid, AttrNumber colId);
 extern PGDLLEXPORT inherit_view_constraints_from_table_hook_type inherit_view_constraints_from_table_hook;
+
+extern ObjectAddress DefineVirtualRelation(RangeVar *relation, List *tlist, bool replace,
+					  						List *options, Query *viewParse);
+
```
We do not directly mark the function as extern instead. Instead declare a new function in view.h/.c and exposed it through that. For example see `SPI_get_depth()`

R4hul04 replied:
Added wrapper function `bbf_define_virtual_relation`

tanscorpio7 commented on src/backend/rewrite/rewriteHandler.c:
```diff
@@ -4431,6 +4431,14 @@ QueryRewrite(Query *parsetree)
+     *
+     * If this is a view with broken rules, try to repair it
+     * using the definition from babelfish_view_def
+     */
```
Let's remove the comments from here.
And an indentation fix `hook)(pa` -> `hook) (pa`

tanscorpio7 commented on src/backend/rewrite/rewriteHandler.c:
```diff
@@ -4403,7 +4404,6 @@ RewriteQuery(Query *parsetree, List *rewrite_events, int orig_rt_length)
 	return rewritten;
 }
 
-
```
Not needed change.

tanscorpio7 commented on src/backend/commands/dropcmds.c:
```diff
@@ -131,6 +133,11 @@ RemoveObjects(DropStmt *stmt)
+
+		if (object_access_hook && (stmt->removeType == OBJECT_FUNCTION) && sql_dialect == SQL_DIALECT_TSQL)
+		{
+			InvokeObjectDropHook(ProcedureRelationId,address.objectId,-1);
```
Indentation: spaces between args.

Commit:
Added wrapper function on DefineVirtualRelation to be used during
babelfish view repair

Signed-off-by: Rahul Parande <rparande@amazon.com>

Commit:
Added back extra line in rewriteHandler.c to avoid merge conflicts

Signed-off-by: Rahul Parande <rparande@amazon.com>
